Fixes for TTree with complex friend hierarchies

vepadulano opened the pull request:
This PR collects the fixes and correlated testing in the latest campaign to fix the CMS DAS MiniAOD analysis framework, the only MiniAOD analysis system supported by the CMS Common Analysis Tools group (see https://indico.cern.ch/event/1330797/papers/5796853/files/13266-CAT_ACAT24_proceeding.pdf)

ferdymercury commented:
> For each of those, we need a corresponding reproducer that fails without the fix and works with the fix. For now those places are not interfering immediately with the code changes needed for this PR, so I would in principle tackle them separately.

Thanks, got it. An easy failures-reproducer of two of the locations: TTreeplayer lines 237 and 620:

change line 83 of tutorials/io/tree/tree109_friend.C with
`auto TF = T->CopyTree("z<10", 0, TTree::kMaxEntries, 2);`
and you'll get an integer overflow. It affects both locations/functions of TTreePlayer. (One should not need to know how many entries the tree has in total even if he wants to skip a few from the beginning).

(For the other locations I do not have a repro.)

pcanal commented on tree/tree/src/TChain.cxx:
```diff
@@ -804,12 +805,17 @@ Long64_t TChain::Draw(const char* varexp, const char* selection,
 
 TBranch* TChain::FindBranch(const char* branchname)
 {
+   std::string name{branchname};
```
Can you add a code comment explaining the net/resulting effect of conditionally stripping the chain's name from the branch name?

pcanal commented on tree/tree/src/TTree.cxx:
```diff
@@ -6122,30 +6122,52 @@ const char* TTree::GetFriendAlias(TTree* tree) const
       return nullptr;
    }
-   if (!fFriends) {
+   TFriendLock lock(const_cast<TTree *>(this), kGetFriendAlias);
```
This is because this could be a chain (and there is no `TChain::GetFriendAlias`, adding it would be another valid implementation choice).  So that early return would now be:
```
if (!friend && GetTree() == this)
	return nullptr;
```

pcanal commented on tree/tree/src/TTree.cxx:
```diff
@@ -6122,30 +6122,52 @@ const char* TTree::GetFriendAlias(TTree* tree) const
-      if (t && t->GetTree() == tree) {
-         return fe->GetName();
+
+   // Then, check if this is a TChain and the current tree has friends
```
We might  not be able to avoid it but in the usual case, this should be redundant.

If this is a chain and it has friends, it will make its underlying tree have the friends too (so same semantic friends).

The case that is non-redundantly covered here is the case where the underlying TTree of a TChain has its own persistent friends.

If this analysis is correct, we ought to clarify it in the code comment (for example to avoid a future maintainer thinking only of the former case and wanting to remove this code)

vepadulano replied:
I believe this is already non-redundant because of the condition
```
innerListOfFriends && innerListOfFriends != fFriends
```
but I'm also adding a comment to make this explicit

vepadulano commented on tree/tree/src/TChain.cxx:
```diff
@@ -804,12 +805,17 @@ Long64_t TChain::Draw(const char* varexp, const char* selection,
 
 TBranch* TChain::FindBranch(const char* branchname)
 {
+   std::string name{branchname};
```
The equivalent in `TTree::FindBranch` is already there https://github.com/root-project/root/blob/4d196eacc4cde31343e5c5705d6e95091d1bef2d/tree/tree/src/TTree.cxx#L4903-L4908 

This is just aligning that logic to TChain as well

vepadulano commented:
@pcanal I addressed all your comments. I also added in the last commit an extension to one of the testing suites that I introduced in an earlier commit, now it considers all cases in which a call to TChain::Scan is accessing one of the friend chains with a name of chain that may be different than the name of tree it holds. ATOW, this PR introduces the following amount of tests from `gh20033.cxx`
```
586: [----------] Global test environment tear-down
586: [==========] 4080 tests from 3 test suites ran. (17421 ms total)
586: [  PASSED  ] 4064 tests.
586: [  SKIPPED ] 16 tests, listed below:
```

+ 4 from `regressions.cxx`

vepadulano commented:
As discussed privately with @pcanal, I have reorganised the commit history of this PR to have all the code changes at the beginning, and separate every type of new test scenario added in a different commit, appended after the changes. Another PR has been opened just with the test commits to see the failures in isolation --> https://github.com/root-project/root/pull/20263
